### PR TITLE
Add DISTINCT query support

### PR DESCRIPTION
**Pull request by southpolesteve**
Should be merged after #306

### CODE REVIEW

File: src/queryExecutionContext/EndpointComponent/OrderedDistinctEndpointComponent.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { sha1 } from "crypto-hash";
```
Comment (southpolesteve) [inline src/queryExecutionContext/EndpointComponent/OrderedDistinctEndpointComponent.ts]: @bchong95 @christopheranderson I ended up going with sha1 (160 bits) for this hash as it is basically zero bundle cost. The browser and node both have a built-in sha1 hash and the `crypto-hash` package handles loading the correct one in each environment. Happy to switch this over to murmurhash once that returns, but this seems like the simplest thing for now.

Comment (bchong95) [reply src/queryExecutionContext/EndpointComponent/OrderedDistinctEndpointComponent.ts]: Im good with sha1, since the hash doesn't need to be crypographically secure:

https://en.wikipedia.org/wiki/SHA-1#SHAttered_%E2%80%93_first_public_collision

Only vulnerablity I can think of is if someone using CosmosDB allows a user to insert user data and they generate a collision to make another person's document not show up in the distinct result. But as the wiki says it would like 1100 years of CPU time, which I am not too worried about.

Comment (southpolesteve) [reply src/queryExecutionContext/EndpointComponent/OrderedDistinctEndpointComponent.ts]: I'll put something on the calendar for 3119

File: src/queryExecutionContext/EndpointComponent/UnorderedDistinctEndpointComponent.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { sha1 } from "crypto-hash";
+import stableStringify from "fast-json-stable-stringify";
+import { Response } from "../../request";
+import { ExecutionContext } from "../ExecutionContext";
+
+/** @hidden */
+export class UnorderedDistinctEndpointComponent implements ExecutionContext {
```
Comment (bchong95) [inline src/queryExecutionContext/EndpointComponent/UnorderedDistinctEndpointComponent.ts]: Can we make these classes implement some `DistinctEndpointComponentBase` class and make these private nested. The base class can then have a factory create method that take the `DistinctType` and `new`s up the appropriate ExecutionContext.

File: src/queryExecutionContext/pipelinedQueryExecutionContext.ts [hunk 3]
```diff
@@ -72,6 +74,15 @@ export class PipelinedQueryExecutionContext implements ExecutionContext {
     if (typeof limit === "number" && typeof offset === "number") {
       this.endpoint = new OffsetLimitEndpointComponent(this.endpoint, offset, limit);
     }
+
+    // If distinct then add that to the pipeline
+    const distinctType = partitionedQueryExecutionInfo.queryInfo.distinctType;
```
Comment (bchong95) [inline src/queryExecutionContext/pipelinedQueryExecutionContext.ts]: Can we move this to the factory method of a distinct document query execution component base class. Ordered vs unordered is an implementation detail for `DISTINCT` and the pipeline should not really concern itself with it. If the distinct component wanted to it can new up an unordered componenet for every query if it wanted to.

File: test/integration/crossPartition.spec.ts [hunk 4]
```diff
@@ -340,6 +341,34 @@ describe("Cross Partition", function() {
       await executeQueryAndValidateResults({ query, options, expectedOrderIds: expectedOrderedIds });
     });
 
+    it("Validate DISTINCT Query", async function() {
+      // simple order by query in string format
+      const query = "SELECT DISTINCT VALUE r.spam3 FROM root r";
+      const options = {
+        enableCrossPartitionQuery: true,
+        maxItemCount: 2
```
Comment (bchong95) [inline test/integration/crossPartition.spec.ts]: Please at parallelism to test any possible race conditions.

File: test/integration/crossPartition.spec.ts [hunk 5]
```diff
@@ -340,6 +341,34 @@ describe("Cross Partition", function() {
...
+      };
+
+      const expectedOrderedIds = Array.from(new Set(documentDefinitions.map(r => r["spam3"])));
```
Comment (bchong95) [inline test/integration/crossPartition.spec.ts]: Is `Set` ordered? If not then the order is undefined and this test will pass by chance.

File: test/integration/crossPartition.spec.ts [hunk 6]
```diff
@@ -340,6 +341,34 @@ describe("Cross Partition", function() {
...
+
+    it("Validate DISTINCT OrderBy Query", async function() {
+      // simple order by query in string format
+      const query = "SELECT DISTINCT VALUE r.spam3 FROM root r order by r.spam3";
+      const options = {
+        enableCrossPartitionQuery: true,
+        maxItemCount: 2
+      };
+
+      const expectedOrderedIds = Array.from(new Set(documentDefinitions.map(r => r["spam3"])));
```
Comment (bchong95) [inline test/integration/crossPartition.spec.ts]: The test code seems the same here as the unordered variant. Shouldn't the results differ?

File: test/integration/crossPartition.spec.ts [hunk 7]
```diff
@@ -340,6 +341,34 @@ describe("Cross Partition", function() {
...
+      // validates the results size and order
+      await executeQueryAndValidateResults({ query, options, expectedOrderIds: expectedOrderedIds });
+    });
+
+    it("Validate DISTINCT OrderBy Query", async function() {
+      // simple order by query in string format
+      const query = "SELECT DISTINCT VALUE r.spam3 FROM root r order by r.spam3";
+      const options = {
+        enableCrossPartitionQuery: true,
+        maxItemCount: 2
```
Comment (bchong95) [inline test/integration/crossPartition.spec.ts]: Please add tests that vary the `maxItemCount`. `maxItemCount` = 1 and collectionSize are special edge cases.

`maxItemCount` = 1 means the backend will essentially do no work and client is responsible for doing all the DISTINCT
`maxItemCount` = collectionSize means the backend will essential do all the work and the clients job is just not to lose any documents.

File: test/integration/crossPartition.spec.ts (hunk 6 above)
Comment (southpolesteve) [reply test/integration/crossPartition.spec.ts]: Yep! It was actually flaky. Sometimes they matched. Fixed

File: test/integration/crossPartition.spec.ts (hunk 7 above)
Comment (southpolesteve) [reply test/integration/crossPartition.spec.ts]: Done

File: test/integration/crossPartition.spec.ts (hunk 5 above)
Comment (southpolesteve) [reply test/integration/crossPartition.spec.ts]: Removed `Set` here

File: src/queryExecutionContext/pipelinedQueryExecutionContext.ts (hunk 3 above)
Comment (southpolesteve) [reply src/queryExecutionContext/pipelinedQueryExecutionContext.ts]: Prefer to keep this explicit for now. Trying to come back and refactor the query pipeline later and would like to have less inheritance and complexity

File: src/queryExecutionContext/EndpointComponent/UnorderedDistinctEndpointComponent.ts (hunk 2 above)
Comment (southpolesteve) [reply src/queryExecutionContext/EndpointComponent/UnorderedDistinctEndpointComponent.ts]: Same as other comments, prefer to keep this explicit for now. Trying to come back and refactor the query pipeline later and would like to have less inheritance and complexity
